src/bluetooth/qbluetoothservicediscoveryagent_bluez.cpp

```diff
@@ -130,45 +129,75 @@ void QBluetoothServiceDiscoveryAgentPrivate::stop()
     emit q->canceled();
 }
 
-void QBluetoothServiceDiscoveryAgentPrivate::_q_createdDevice(QDBusPendingCallWatcher *watcher)
+void QBluetoothServiceDiscoveryAgentPrivate::_q_foundDevice(QDBusPendingCallWatcher *watcher)
 {
-    if (!adapter)
+    if (!adapter) {
+        watcher->deleteLater();
         return;
+    }
 
     Q_Q(QBluetoothServiceDiscoveryAgent);
 
     const QBluetoothAddress &address = watcher->property("_q_BTaddress").value<QBluetoothAddress>();
 
-    qCDebug(QT_BT_BLUEZ) << Q_FUNC_INFO << "created" << address.toString();
+    qCDebug(QT_BT_BLUEZ) << Q_FUNC_INFO << "found" << address.toString();
 
     QDBusPendingReply<QDBusObjectPath> deviceObjectPath = *watcher;
+    watcher->deleteLater();
     if (deviceObjectPath.isError()) {
-        if (deviceObjectPath.error().name() != QLatin1String("org.bluez.Error.AlreadyExists")) {
+        if (deviceObjectPath.error().name() != QLatin1String("org.bluez.Error.DoesNotExist")) {
+            qCDebug(QT_BT_BLUEZ) << "Find device failed Error: " << error << deviceObjectPath.error().name();
             delete adapter;
             adapter = 0;
             _q_serviceDiscoveryFinished();
-            qCDebug(QT_BT_BLUEZ) << "Create device failed Error: " << error << deviceObjectPath.error().name();
             return;
         }
 
-        deviceObjectPath = adapter->FindDevice(address.toString());
-        deviceObjectPath.waitForFinished();
-        if (deviceObjectPath.isError()) {
+        deviceObjectPath = adapter->CreateDevice(address.toString());
+        watcher = new QDBusPendingCallWatcher(deviceObjectPath, q);
+        watcher->setProperty("_q_BTaddress", QVariant::fromValue(address));
+        QObject::connect(watcher, SIGNAL(finished(QDBusPendingCallWatcher*)),
+                         q, SLOT(_q_createdDevice(QDBusPendingCallWatcher*)));
+        return;
+    }
+
+    qCDebug(QT_BT_BLUEZ) << Q_FUNC_INFO << "path" << deviceObjectPath.value().path();
+    discoverServices(deviceObjectPath.value().path());
+}
+
+void QBluetoothServiceDiscoveryAgentPrivate::_q_createdDevice(QDBusPendingCallWatcher *watcher)
+{
+    if (!adapter) {
+        watcher->deleteLater();
+        return;
+    }
+
+    const QBluetoothAddress &address = watcher->property("_q_BTaddress").value<QBluetoothAddress>();
+
+    qCDebug(QT_BT_BLUEZ) << Q_FUNC_INFO << "created" << address.toString();
+
+    QDBusPendingReply<QDBusObjectPath> deviceObjectPath = *watcher;
+    watcher->deleteLater();
+    if (deviceObjectPath.isError()) {
+        if (deviceObjectPath.error().name() != QLatin1String("org.bluez.Error.AlreadyExists")) {
+            qCDebug(QT_BT_BLUEZ) << "Create device failed Error: " << error << deviceObjectPath.error().name();
             delete adapter;
             adapter = 0;
-            if (singleDevice) {
```

Comment:
As pointed out on IRC, this removal isn't quite right. Error state needs to be reported up.